appsec-kit-v7/src/main/java/com/vaadin/appsec/v7/service/NotificationInitializer.java

```diff
@@ -56,32 +63,37 @@ static void serviceInit(VaadinService vaadinService) {
         if (isDebugMode(vaadinService)) {
             vaadinService.addSessionInitListener(e -> {
                 e.getSession().addUIProvider(new AppSecUIProvider());
-                createNotificationThread(e.getSession()).start();
+                startNotificationTask(e.getSession());
             });
             LOGGER.info("NotificationInitListener initialized.");
         }
     }
 
-    private static Thread createNotificationThread(
-            final VaadinSession session) {
-        return new Thread(() -> {
-            try {
-                LOGGER.info(
-                        "NotificationInitListener notification thread initialized.");
-
-                Thread.sleep(NOTIFICATION_CHECK_INTERVAL);
-
-                while (isSessionOpen(session)) {
-                    session.access(() -> {
-                        session.getUIs()
-                                .forEach(ui -> notifyUiIfNeeded(session, ui));
-                    });
-                    Thread.sleep(NOTIFICATION_CHECK_INTERVAL);
-                }
-            } catch (InterruptedException e) {
-                // NOP
-            }
-        });
+    private static void startNotificationTask(final VaadinSession session) {
+        final String sessionId = session.getSession().getId();
+        final ScheduledFuture<?> scheduledNotificationCheck = AppSecService
+                .getInstance().getConfiguration().getTaskExecutor()
+                .scheduleAtFixedRate(() -> {
+                    LOGGER.info(
+                            "NotificationInitListener notification thread initialized.");
+
+                    if (isSessionOpen(session)) {
+                        session.access(() -> {
+                            session.getUIs().forEach(
+                                    ui -> notifyUiIfNeeded(session, ui));
+                        });
+                    } else {
+                        ScheduledFuture<?> thisCheck = scheduledChecks
+                                .get(sessionId);
+                        if (thisCheck != null) {
+                            thisCheck.cancel(true);
+                            scheduledChecks.remove(sessionId);
+                        }
+                    }
+                }, NOTIFICATION_CHECK_INTERVAL, NOTIFICATION_CHECK_INTERVAL,
+                        TimeUnit.MILLISECONDS);
+        scheduledChecks.put(session.getSession().getId(),
```

Comment:
Could be the `sessionId` used here also?

Reply:
Yes, thanks!